### PR TITLE
[JS] Add Middleware Documentation

### PR DESCRIPTION
**Pull request by FrankEssenberger**
Relates to: https://github.com/SAP/cloud-sdk-backlog/issues/669

This PR adds some documentation on the middleware in general. The resilience section needs to be updated as well. I will do this once the PRs are merged and the API is 100% solid.

### CODE REVIEW

File: docs-js/features/middleware.mdx [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+- The `fn` handed in is the original HTTP request from the SAP Cloud SDK.
+- The `context` provides information on the request like URL, headers and HTTP method.
+- The `skipNext` will exit all following middlewares
+- The `MiddlewareOut` is the new function including the additional logic.
+
+This seems abstract in the beginning, but the example in the next section will bring some concreteness to the abstract topic.
+
+## Example Middleware
+
+As a simple example assume you want to ensure that your response data does not contain only uppercase characters.
```
Comment (fwilhe) [inline docs-js/features/middleware.mdx:50]:
```suggestion
As a simple example assume you want to ensure that your response data does contain only uppercase characters.
```

i think the 'not' is wrong here

File: docs-js/features/middleware.mdx [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+response = await executeHttpRequest(httpsDestination, {
+  middleware: [uppercaseMiddleware, addFortyTwoMiddleware],
+  method: 'get'
+});
+
+console.log(response.data); // 'THE ANSWER TO EVERYTHING IS: Forty-Two.
+```
+
+In the context of resilience the order of middlewares becomes crucial.
```
Comment (deekshas8) [inline docs-js/features/middleware.mdx:95]:
```suggestion
In the context of resilience, the order of middlewares becomes crucial.
```

File: docs-js/features/middleware.mdx [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+A timeout should be added first, then a circuit breaker followed by a retry.
+The SAP Cloud SDK provides default resilience middlewares so that you do not have to worry about the details in most cases.
+You find detailed information on the resilience topic in a [dedicated documentation](../guides/resilience).
```
Comment (deekshas8) [inline docs-js/features/middleware.mdx:98]:
```suggestion
You can find detailed information on the resilience topic in a [dedicated documentation](../guides/resilience).
```

File: docs-js/features/openapi/execute-openapi-request.mdx [hunk 4]
```diff
@@ -104,25 +104,16 @@ MyEntity.requestBuilder()
...
-</TabItem>
-</Tabs>
-
-The value is in milliseconds and the default value is 10000 (10 seconds)
-This timeout applies to the request send to the target system given in the destination.
-If a CSRF token is fetched from the target system the timeout applies also to this call.
-Note that there is also a second timeout option on the destination fetch options.
-This relates to the calls made to SAP BTP services like destination-service or XSUAA.
+A typical use case for middlewares is to add [resilience](../../guides/resilience) to your request.
+The concept is more general and can be used to add [arbitrary enhancements of the request](../middleware).
```
Comment (deekshas8) [inline docs-js/features/openapi/execute-openapi-request.mdx:116]:
```suggestion
The concept is more general and can be used to add [arbitrary enhancements to the request](../middleware).
```

File: docs-js/features/odata/execute-odata-request.mdx [hunk 5]
```diff
@@ -280,24 +280,21 @@ where:
...
   .execute(destination);
 ```
 
-The value is in milliseconds and the default value is 10000 (10 seconds)
-This timeout applies to the request sent to the target system given in the destination.
-If a cross-site request forgery (CSRF) token is fetched from the target system the timeout applies also to this call.
-Note that there is also a second timeout option on the destination fetch options.
-This relates to the calls made to SAP BTP services like destination service or XSUAA.
+A typical use case for middlewares is to add [resilience](../../guides/resilience) to your request.
+The concept is more general and can be used to add [arbitrary enhancements of the request](../middleware).
```
Comment (deekshas8) [inline docs-js/features/odata/execute-odata-request.mdx:297]:
```suggestion
The concept is more general and can be used to add [arbitrary enhancements to the request](../middleware).
```

File: docs-js/features/middleware.mdx [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+A middleware doing that would look like:
+
+```ts
+const uppercaseMiddleware: Middleware = options => async () => {
+  const response = await options.fn();
+  return response.data.toUpperCase();
+};
+```
+
+and you would add it via the `middleware` property to your request:
```
Comment (deekshas8) [inline docs-js/features/middleware.mdx:60]:
```suggestion
Add it to the request via the `middleware` property as shown below:
```

File: docs-js/features/middleware.mdx [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+
+## Middleware Concept
+
+The main purpose of the SAP Cloud SDK is to execute asynchronous HTTP requests.
+This can either happen via a generated [OData](./odata/execute-request#setting-middlewares) or [OpenAPI](./openapi/execute-request#setting-middlewares) client or the [http client](./connectivity/http-client#setting-middlewares).
+Either way, the final result after fetching destinations and doing various other things is an [axios](https://github.com/axios/axios) request.
+Sometimes you want to adjust the way the SAP Cloud SDK executes the axios request.
+
+The middleware adds a very flexible way to add custom functionality to the request execution.
+The API for the middleware looks like:
```
Comment (deekshas8) [inline docs-js/features/middleware.mdx:24]: [pp] The API for the middleware is as follows:

[req] complete sentence.

File: docs-js/features/odata/execute-odata-request.mdx [hunk 8]
```diff
@@ -280,24 +280,21 @@ where:
...
   .execute(destination);
 ```
 
-The value is in milliseconds and the default value is 10000 (10 seconds)
-This timeout applies to the request sent to the target system given in the destination.
-If a cross-site request forgery (CSRF) token is fetched from the target system the timeout applies also to this call.
-Note that there is also a second timeout option on the destination fetch options.
-This relates to the calls made to SAP BTP services like destination service or XSUAA.
+A typical use case for middlewares is to add [resilience](../../guides/resilience) to your request.
+The concept is more general and can be used to add [arbitrary enhancements to the request](../middleware).
```
Comment (deekshas8) [inline docs-js/features/odata/execute-odata-request.mdx:297]: [pp] From my understanding, it says here that middleware is a general concept and one use case is adding resilience.

In that case, I would switch the sentences and rewrite:

Middleware is a general concept used to add [arbitrary enhancements to a request](../middleware).
A typical use case is to also add [resilience](../../guides/resilience) to requests.

**Review by deekshas8 — COMMENTED** [review]
Many of the comments are [pp]. Feel free to consider as you like.

File: docs-js/features/middleware.mdx [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+  - sap
+  - cloud
+  - sdk
+  - JavaScript
+  - TypeScript
+  - resilience
+  - middleware
+---
+
+## Middleware Concept
```
[the remark this answers is not in the document]
Comment (FrankEssenberger) [reply docs-js/features/middleware.mdx:16]: yes I like this

File: docs-js/features/middleware.mdx [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+
+The middleware receives a `MiddlewareIn` option and returns a `MiddlewareOut`.
+Let's investigate the different properties:
+
+- The `fn` handed in is the original HTTP request from the SAP Cloud SDK.
+- The `context` provides information on the request like URL, headers and HTTP method.
+- The `skipNext` will exit all following middlewares
+- The `MiddlewareOut` is the new function including the additional logic.
+
+This seems abstract in the beginning, but the example in the next section will bring some concreteness to the abstract topic.
```
[the remark this answers is not in the document]
Comment (FrankEssenberger) [reply docs-js/features/middleware.mdx:46]: like it

File: docs-js/features/connectivity/generic-http-client.mdx [hunk 11]
```diff
@@ -183,6 +183,20 @@ const response = await executeHttpRequest(
...
+You can specify middlewares for a request via the `middleware` property of the request config
+
+```ts
+const response = await executeHttpRequest(httpsDestination, {
+  middleware: [myMiddleware],
+  method: 'get'
+});
+```
+
+A typical use case for middlewares is to add [resilience](../../guides/resilience) to your request.
```
Comment (deekshas8) [inline docs-js/features/connectivity/generic-http-client.mdx:197]:
```suggestion
Middleware is a general concept used to add [arbitrary enhancements to the request](../middleware).
A typical use case is to also add [resilience](../../guides/resilience) to requests.
```

File: docs-js/features/connectivity/generic-http-client.mdx [hunk 12]
```diff
@@ -183,6 +183,20 @@ const response = await executeHttpRequest(
...
+The concept is more general and can be used to add [arbitrary enhancements to the request](../middleware).
```
Comment (deekshas8) [inline docs-js/features/connectivity/generic-http-client.mdx:198]:
```suggestion
```

**Review by deekshas8 — APPROVED** [review]
LGTM.
Please adjust the sentence in http-client as well.

**Pull request merged**